File: 2011/B/Haskell/solve.hs
-- -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
--
-- File Name : solve.hs
--
-- Purpose :
--
-- Creation Date : 06-09-2011
--
-- Last Modified : Thu 15 Sep 2011 07:18:32 PM EEST
--
-- Created By : Greg Liras <gregliras@gmail.com>
--
--_._._._._._._._._._._._._._._._._._._._._.


import List
import Data.Set
 
import Data.List (intercalate)




  





rInt x = read x :: Integer


getNM :: IO [Integer]
getNM =
  do
    inp <- getLine
    return (List.map rInt (Main.split inp ' '))
getLines   :: Integer->Integer ->[String]-> IO [String]
getLines n maX acc
  | n <= maX =
    do
      inp <- getLine
      getLines (n+1) maX (inp:acc)
  | otherwise = return (reverse acc)


instance Show MyData where
  show (Words strings integer) = show strings ++ " , " ++ show integer

data MyData = Words [String] Int


buildDataList ::[[String]]->[MyData]
buildDataList [] = []
buildDataList (wordList:rList) = Words wordList 0 : buildDataList rList
  
    
filterByLength x ls = List.filter (\n -> length n == x) ls

filterByElemIndices ::String->[String]->Char->[String]
filterByElemIndices word wordlist letter = 
  let
    indices = elemIndices letter word
  in
    List.filter (\x -> (elemIndices letter x) == indices) wordlist

filterByFirstOccurance :: [String]->[String]->String
filterByFirstOccurance (dict:ionary) answers =
  if elem dict answers
  then dict
  else filterByFirstOccurance ionary answers

filterByMaxInt answers =
  let
    (sstr,iint) = unzip answers
  in
    let
      (answer,_) = unzip (List.filter (\(x,y) -> y==(maximum iint)) answers)
    in
      answer

getLengthSists x ls = getLengthSistsH x ls []

getLengthSistsH [] _ acc = reverse acc
getLengthSistsH (x:xs) ls acc = getLengthSistsH xs ls ((filterByLength x ls):acc)

lengthSet :: [[a]]-> [Int]
lengthSet ls = toList (fromList (List.map length ls))

split :: String -> Char -> [String]
split [] delim = [""]
split (c:cs) delim
   | c == delim = "" : rest
   | otherwise = (c : head rest) : tail rest
         where
                rest = Main.split cs delim


groupByLength ls = 
  let
    lengthS = lengthSet ls
  in
    getLengthSists lengthS ls



findMaxWordFast dictionary lettersList = 
  let
    groupedByLength = groupByLength dictionary
  in
    let
      allAnswers = List.map (filterByFirstOccurance dictionary) (findEveryThing [(countSteps (buildDataList groupedByLength) x)  | x <- lettersList])
    in
      allAnswers --[filterByFirstOccurance x dictionary | x <- allAnswers]



findEveryThing :: [[MyData]]-> [[String]]
findEveryThing myDataLists =
  List.map (\x -> let max = getMaxCounter x 0 in filterByMaxCounter x max []) myDataLists


filterByMaxCounter :: [MyData] -> Int -> [[String]] -> [String]
filterByMaxCounter [] _ acc = concat acc
filterByMaxCounter ((Words string counter):rlist) max acc = 
  if (counter == max)
    then filterByMaxCounter rlist max (string:acc)
    else filterByMaxCounter rlist max acc

getMaxCounter [] max = max
getMaxCounter ((Words _ counter):rlist) max =
  if (max < counter) 
    then getMaxCounter rlist counter
    else getMaxCounter rlist max



countSteps ::[MyData]->String->[MyData]
countSteps wordListList [] = wordListList
countSteps wordListList (letter:s) = 
  let
    nextWordListList = concat [groupByElemIndices letter wordlist | wordlist <- wordListList]
  in
    countSteps nextWordListList s

        
getWordListsFromTuples tupleWordList = 
  let
    (a,b) = unzip tupleWordList
  in
    a
  

getElemIndicesForAllWords wordList letter= 
  [(x,y) | x <- wordList,let y = elemIndices letter x]




groupByElemIndices :: Char->MyData->[MyData]
groupByElemIndices letter (Words wordList counter) = 
  let
    toupleWordList = getElemIndicesForAllWords wordList letter
  in
    buildNextList (groupByElemIndicesH toupleWordList []) counter





buildNextList :: [[(String,[Int])]]->Int->[MyData]
buildNextList toupleWordLists prevCounter =
  let
    wordListsWithNoLetter = List.map getWordListsFromTuples (List.filter (any (\(x,y) -> y == [] )) toupleWordLists)
    wordListsWithLetter = List.map getWordListsFromTuples (List.filter (any (\(x,y) -> y /= [] ) ) toupleWordLists)
    wasIn = not (List.null wordListsWithLetter)
  in
    if wasIn
    then 
      let 
        nextDataWithout = [Words x (prevCounter+1) | x <- wordListsWithNoLetter]
        nextDataWith = [Words x prevCounter | x <- wordListsWithLetter]
      in
        nextDataWith++nextDataWithout
    else
      [Words x prevCounter | x <- wordListsWithNoLetter]


groupByElemIndicesH ::[(String,[Int])]->[[(String,[Int])]]->[[(String,[Int])]]
groupByElemIndicesH [] acc = acc
groupByElemIndicesH ((a,b):leWordList) acc = 
  let
    first = List.filter (\(x,y) -> y==b) leWordList
    rest = List.filter (\(x,y) -> y/=b) leWordList
  in
    groupByElemIndicesH rest (((a,b):first):acc)




getAll x cases
  | x<= cases =
    do
      nm <- getNM
      dictionary <- getLines 1 (head(nm)) []
      sequences <- getLines 1 (head(tail(nm))) []
      putStrLn ("Case #"++(show x)++": "++(intercalate " "(findMaxWordFast dictionary sequences)))
      getAll (x+1) cases
  | otherwise = return ()
                    
main = 
  do
    cases <- readLn
    --putStrLn (show cases)
    getAll 1 cases

File: 2011/B/C++/solve.cpp
/* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.

* File Name : solve.cpp

* Purpose :

* Creation Date : 11-09-2011

* Last Modified : Sun 11 Sep 2011 08:31:12 PM EEST

* Created By : Greg Liras <gregliras@gmail.com>

_._._._._._._._._._._._._._._._._._._._._.*/

#include <iostream>
#include <cstring>
#include <cstdlib>
#include <list>
#include <vector>


using namespace std;



class mySuperClass
{
  private:
    list<string *> *contents;
    int counter;
    list < vector < int > * > *buildIndices(char c);
    bool letterIn;
  public:
    ~mySuperClass();
    mySuperClass(list<string *> *newContents,int newCounter);
    list<mySuperClass *> *next(char c);
    int getCounter();
    list<string *> *getString();
};



list<mySuperClass *> *buildStartWordListList(list<string *> *words);
list<mySuperClass *> *getNextLevel(list<mySuperClass *> *myWordList,char c);
string filterMaxString(list <string *> *dictionary, list <string *> *maxString);
list <string *> * findMax(list<mySuperClass *> *wordLists);


int main (int argc,char ** argv)
{
  int cases=0;
  cin >> cases;
  for(int i=1;i<=cases;i++)
  {
    int words=0;
    int letterSets=0;
    list<string *> *wordList = new list<string*>();
    list<string *> *letterList = new list<string*>();
    list<mySuperClass *> *wordLists;
    list<mySuperClass *> *currentWordLists;
    list<string *> *maxString;
    cout << "Case #" <<i<<":";
    cin >> words;
    cin >> letterSets;
    string *buff;
    for (int j=0;j<words;j++)
    {
      buff = new string();
      cin >> *buff;
      wordList->push_back(buff);
    }
    for (int j=0;j<letterSets;j++)
    {
      buff = new string();
      cin >> *buff;
      letterList->push_back(buff);
    }
    wordLists = buildStartWordListList(wordList);
    currentWordLists = wordLists;
    
    list<string *>::iterator letterString = letterList->begin();
    for(;letterString!=letterList->end();letterString++)
    {
      for (int j=0;j<26;j++)
      {
        currentWordLists = getNextLevel(currentWordLists,(**letterString)[j]);
      }
      maxString = findMax(currentWordLists);
      cout << " " <<filterMaxString(wordList,maxString);

    }
  }
}


string filterMaxString(list <string *> *dictionary, list <string *> *maxString)
{
  list<string *>::iterator finder = dictionary->begin();
  list<string *>::iterator sndFinder = maxString->begin();
  for(;finder!=dictionary->end();finder++)
  {
    for(;sndFinder!=maxString->end();sndFinder++)
    {
      if ((**finder)==(**sndFinder)) return **finder;
    }
  }
  return "NOTME!!!!";
}

list <string *> * findMax(list<mySuperClass *> *wordLists)
{
  int buf,max = -1;
  list<string *> * maxString;
  list<mySuperClass *>::iterator maxFound=wordLists->begin();
  for(;maxFound!=wordLists->end();maxFound++)
  {
    if((buf=(*maxFound)->getCounter())>max)
    {
      max=buf;
      maxString = (*maxFound)->getString();
    }
  }
  return maxString;
}

list<mySuperClass *> *getNextLevel(list<mySuperClass *> *myWordList,char c)
{
  list<mySuperClass *>::iterator current = myWordList->begin();
  list<mySuperClass *> *nextList = new list<mySuperClass *>();
  for(;current!=myWordList->end();current++)
  {
    list<mySuperClass *> *buff = (*current)->next(c);
    for (list<mySuperClass *>::iterator bf = buff->begin();bf!=buff->end();bf++)
    {
      nextList->push_back(*bf);
    }

  }
  return nextList;
}
list<mySuperClass *> *buildStartWordListList(list<string *> *words)
{
  list<string *>::iterator iter = words->begin();
  list<mySuperClass *> *rList = new list<mySuperClass *>();
  list<string *> *buff;
  for(int i = 1 ; i<=10;i++)
  {
    buff = new list<string *>();
    for(;iter!=words->end();iter++)
    {
      if ((*iter)->length()==i)
      {
        buff->push_back(*iter);
      }
    }
    rList->push_back(new mySuperClass(buff,0));
  }
  return rList;
}

list<string *> * mySuperClass::getString()
{
  return contents;
}

int mySuperClass::getCounter()
{
  return counter;
}
list < vector < int > *> *mySuperClass::buildIndices(char c)
{
  list<string *>::iterator word = contents->begin();
  list < vector < int > * > *indicesList = new list < vector < int >* >;
  vector < int > *buff;
  for (;word!=contents->end();word++)
  {
    buff = new vector<int>();
    for(int i=0;i<(**word).length();i++)
    {
      if (c== (**word)[i]) (*buff).push_back(c);
    }
    if (buff->size()>0) letterIn = true;
    indicesList->push_back(buff);
  }
  return indicesList;
}

mySuperClass::mySuperClass(list<string *> *newContents,int c)
{
  counter = c;
  contents = newContents;
  letterIn = false;
}

list<mySuperClass *> *mySuperClass::next(char c)
{
  list < vector < int > *> *allIndices = buildIndices(c);
  list<string*>::iterator word = contents->begin();
  list<string*>::iterator sndWord = contents->begin();
  list < vector < int > *>::iterator current = allIndices->begin();
  list < vector < int > *>::iterator sndCurrent = allIndices->begin();
  list <mySuperClass *> *rList = new list <mySuperClass *>();
  for (;word!=contents->end(),current!=allIndices->end();word++,current++)
  {
    list<string*> *buffer = new list<string*>();
    for(;sndWord!=contents->end() , sndCurrent!=allIndices->end();sndWord++,sndCurrent++)
    {
      if ((**current)==(**sndCurrent)) buffer->push_back((*sndWord));
    }
    if (((*current)->size()==0)&&letterIn)
    {
      rList->push_back(new mySuperClass(buffer,counter+1));
    }
    else
    {
      rList->push_back(new mySuperClass(buffer,counter));
    }
  }
  return rList;
}

